src/v/kafka/server/handlers/produce.cc

```diff
@@ -348,9 +356,18 @@ produce_topic(produce_ctx& octx, produce_request::topic& topic) {
             continue;
         }
 
+        if (materialized) {
+            partitions_dispatched.push_back(ss::now());
+            partitions_produced.push_back(
+              ss::make_ready_future<produce_response::partition>(
+                produce_response::partition{
+                  .partition_index = part.partition_index,
+                  .error_code = error_code::invalid_topic_exception}));
```

Comment:
i think we could choose a better error message here. looking at kafka source it seems this is mostly used in cases where the client _could_ change something to get things working (e.g. their topic name is invalid). but in this case it is the functionality being rejected. one option might be use treat this like an authorization error. there is one called policy_violation, etc..

Reply:
@twmb do you have any input here? Were kind of always wondering what kind of errors to return to the clients in these situations where we are introducing new behavior while maintaining the same kafka API.

Reply:
What problem is happening here? Is it retriable for the client, or fatal?

POLICY_VIOLATION sounds like a good error. From the client side, it is a non-retriable error. (fwiw, [this](https://pkg.go.dev/github.com/twmb/franz-go/pkg/kerr) is what I use to check errors, but this just mirrors Kafka's Errors.java file in a way).

Reply:
The client should just understand that this is something that will not work - publishing to a materialized topic.

Reply:
@twmb 

> What problem is happening here? Is it retriable for the client, or fatal?

it is not retriable. it's effectively the equivalent of trying to produce to a read-only topic.

Reply:
Policy violation looks to be the most accurate error at the moment, I couldn't see anything better.